x/ibc/03-connection/keeper/handshake.go

```diff
@@ -0,0 +1,242 @@
+	}
+
+	_, found = k.GetConnection(ctx, connectionID)
+	if found {
+		return sdkerrors.Wrap(types.ErrConnectionExists(k.codespace), "cannot relay connection attempt")
+	}
+
+	if !checkVersion(version, counterpartyVersions[0]) {
+		return errors.New("versions don't match") // TODO: sdk.Error
+	}
+
+	connection.State = types.TRYOPEN
+	err = k.addConnectionToClient(ctx, clientID, connectionID)
+	if err != nil {
+		return sdkerrors.Wrap(err, "cannot relay connection attempt")
+	}
+
+	k.SetConnection(ctx, connectionID, connection)
+	k.Logger(ctx).Info(fmt.Sprintf("connection %s state updated: NONE -> TRYOPEN ", connectionID))
+	return nil
+}
+
+// ConnOpenAck relays acceptance of a connection open attempt from chain B back
+// to chain A (this code is executed on chain A).
+func (k Keeper) ConnOpenAck(
+	ctx sdk.Context,
+	connectionID string,
+	version string,
+	proofTry ics23.Proof,
+	proofHeight uint64,
+	consensusHeight uint64,
+) error {
+	// TODO: validateConnectionIdentifier(identifier)
+	if consensusHeight > uint64(ctx.BlockHeight()) {
+		return errors.New("invalid consensus height") // TODO: sdk.Error
+	}
+
+	connection, found := k.GetConnection(ctx, connectionID)
+	if !found {
+		return sdkerrors.Wrap(types.ErrConnectionNotFound(k.codespace), "cannot relay ACK of open attempt")
+	}
+
+	if connection.State != types.INIT {
+		return errors.New("connection is in a non valid state") // TODO: sdk.Error
+	}
+
+	if !checkVersion(connection.Versions[0], version) {
+		return errors.New("versions don't match") // TODO: sdk.Error
+	}
+
+	expectedConsensusState, found := k.clientKeeper.GetConsensusState(ctx, connection.ClientID)
+	if !found {
+		return errors.New("client consensus state not found") // TODO: use ICS02 error
+	}
+
+	prefix := k.clientKeeper.GetCommitmentPath()
+	expectedCounterparty := types.NewCounterparty(connection.ClientID, connectionID, prefix)
+	expectedConn := types.NewConnectionEnd(connection.ClientID, expectedCounterparty, []string{version})
+	expectedConn.State = types.TRYOPEN
+
+	expConnBz, err := k.cdc.MarshalBinaryLengthPrefixed(expectedConn)
+	if err != nil {
+		return err
+	}
+
+	ok := k.VerifyMembership(
+		ctx, connection, proofHeight, proofTry,
+		types.ConnectionPath(connection.Counterparty.ConnectionID), expConnBz,
+	)
+	if !ok {
+		return errors.New("couldn't verify connection membership on counterparty's client") // TODO: sdk.Error
+	}
+
+	expConsStateBz, err := k.cdc.MarshalBinaryLengthPrefixed(expectedConsensusState)
+	if err != nil {
+		return err
+	}
+
+	ok = k.VerifyMembership(
+		ctx, connection, proofHeight, proofTry,
+		ics02types.ConsensusStatePath(connection.Counterparty.ClientID), expConsStateBz,
+	)
+	if !ok {
+		return errors.New("couldn't verify consensus state membership on counterparty's client") // TODO: sdk.Error
+	}
+
+	connection.State = types.OPEN
+
+	// abort if version is the last one
+	// TODO: what does this suppose to mean ?
```

Comment:
What is this TODO asking for?

Reply:
What is this code doing? I don't understand the intent. `checkVersion` should be sufficient.